diff --git a/include/uapi/linux/if_pbb.h b/include/uapi/linux/if_pbb.h
index 7d2995f..090132f 100644
--- a/include/uapi/linux/if_pbb.h
+++ b/include/uapi/linux/if_pbb.h
@@ -29,6 +29,7 @@ enum {
 	IFLA_PBB_I_ISID_INFO_RANGE_END,
 	IFLA_PBB_I_ISID_TYPE,
 	IFLA_PBB_I_KEY_INFO_MAP_VERIFY,
+	IFLA_PBB_I_KEY_INFO_MAP_DEL,
 	IFLA_PBB_I_MAX,
 };
 
@@ -53,6 +54,7 @@ enum {
 	IFLA_PBB_B_VID_INFO_RANGE_BEGIN,
 	IFLA_PBB_B_VID_INFO_RANGE_END,
 	IFLA_PBB_B_KEY_INFO_MAP_VERIFY,
+	IFLA_PBB_B_KEY_INFO_MAP_DEL,
 	IFLA_PBB_B_MAX,
 };
 
diff --git a/ip/link_pbb.c b/ip/link_pbb.c
index 96e6b38..a2fd542 100644
--- a/ip/link_pbb.c
+++ b/ip/link_pbb.c
@@ -147,7 +147,7 @@ static int pbbb_parse_opt(struct link_util *lu, int argc, char **argv,
                        (n->nlmsg_flags & NLM_F_CREATE));
 	__u32 i_sid_info_range_begin = 0, i_sid_info_range_end = 0;
 	int  b_vid_info_range_begin = -1, b_vid_info_range_end = -1;
-	__u8 mode = PBB_B_MODE_MAX, key_info_map_verify = 0;
+	__u8 mode = PBB_B_MODE_MAX, key_info_map_verify = 0, key_info_map_del = 0;
 
 	fprintf(stderr, "Invoking %s with type:0x%x flags:0x%x *argv:%s\n", __FUNCTION__, n->nlmsg_type, n->nlmsg_flags, *argv);
 
@@ -184,6 +184,10 @@ static int pbbb_parse_opt(struct link_util *lu, int argc, char **argv,
 			key_info_map_verify = 1;
 
 			addattr_l(n, 1024, IFLA_PBB_B_KEY_INFO_MAP_VERIFY, &key_info_map_verify, sizeof(__u8));
+		} else if (strcmp(*argv, "key-info-map-del") == 0) {
+			key_info_map_del = 1;
+
+			addattr_l(n, 1024, IFLA_PBB_B_KEY_INFO_MAP_DEL, &key_info_map_del, sizeof(__u8));
 		} else if (matches(*argv, "i-sid") == 0) {
 			char *p = NULL;
 
@@ -265,7 +269,7 @@ static int pbbb_parse_opt(struct link_util *lu, int argc, char **argv,
 				return -1;
 		}
 
-		if ((mode == PBB_B_MODE_MAX) && (key_info_map_verify == 0)) {
+		if ((mode == PBB_B_MODE_MAX) && (key_info_map_verify == 0) && (key_info_map_del == 0)) {
 			return pbbb_b_mode_none_arg();
 		}
 
@@ -290,7 +294,7 @@ static int pbbi_parse_opt(struct link_util *lu, int argc, char **argv,
 	int c_vid_info_range_begin = -1, c_vid_info_range_end = -1;
 	int s_vid_info_range_begin = -1, s_vid_info_range_end = -1;
 	__u32 i_sid_info_range_begin = 0, i_sid_info_range_end = 0;
-	__u8 c_vid_info_action = PBB_I_VID_INFO_ACTION_MAX, s_vid_info_action = PBB_I_VID_INFO_ACTION_MAX, i_sid_type = PBB_I_ISID_TYPE_MAX, key_info_map_verify = 0;
+	__u8 c_vid_info_action = PBB_I_VID_INFO_ACTION_MAX, s_vid_info_action = PBB_I_VID_INFO_ACTION_MAX, i_sid_type = PBB_I_ISID_TYPE_MAX, key_info_map_verify = 0, key_info_map_del = 0;
 
 	fprintf(stderr, "Invoking %s with type:0x%x flags:0x%x\n", __FUNCTION__, n->nlmsg_type, n->nlmsg_flags);
 
@@ -312,6 +316,10 @@ static int pbbi_parse_opt(struct link_util *lu, int argc, char **argv,
 			key_info_map_verify = 1;
 
 			addattr_l(n, 1024, IFLA_PBB_I_KEY_INFO_MAP_VERIFY, &key_info_map_verify, sizeof(__u8));
+		} else if (strcmp(*argv, "key-info-map-del") == 0) {
+			key_info_map_del = 1;
+
+			addattr_l(n, 1024, IFLA_PBB_I_KEY_INFO_MAP_DEL, &key_info_map_del, sizeof(__u8));
 		} else if (matches(*argv, "c-vid") == 0) {
 			char *p = NULL;
 
