diff --git a/include/uapi/linux/if_pbb.h b/include/uapi/linux/if_pbb.h
index 7d2995f..090132f 100644
--- a/include/uapi/linux/if_pbb.h
+++ b/include/uapi/linux/if_pbb.h
@@ -29,6 +29,7 @@ enum {
 	IFLA_PBB_I_ISID_INFO_RANGE_END,
 	IFLA_PBB_I_ISID_TYPE,
 	IFLA_PBB_I_KEY_INFO_MAP_VERIFY,
+	IFLA_PBB_I_KEY_INFO_MAP_DEL,
 	IFLA_PBB_I_MAX,
 };
 
@@ -53,6 +54,7 @@ enum {
 	IFLA_PBB_B_VID_INFO_RANGE_BEGIN,
 	IFLA_PBB_B_VID_INFO_RANGE_END,
 	IFLA_PBB_B_KEY_INFO_MAP_VERIFY,
+	IFLA_PBB_B_KEY_INFO_MAP_DEL,
 	IFLA_PBB_B_MAX,
 };
 
diff --git a/ip/Makefile b/ip/Makefile
index 2ae9df8..91f7bd9 100644
--- a/ip/Makefile
+++ b/ip/Makefile
@@ -5,7 +5,7 @@ IPOBJ=ip.o ipaddress.o ipaddrlabel.o iproute.o iprule.o ipnetns.o \
     ipxfrm.o xfrm_state.o xfrm_policy.o xfrm_monitor.o iplink_dummy.o \
     iplink_ifb.o iplink_nlmon.o iplink_team.o iplink_vcan.o iplink_vxcan.o \
     iplink_vlan.o link_veth.o link_gre.o iplink_can.o iplink_xdp.o \
-    iplink_macvlan.o ipl2tp.o link_vti.o link_vti6.o link_xfrm.o \
+    iplink_macvlan.o ipl2tp.o link_vti.o link_vti6.o link_xfrm.o link_pbb.o \
     iplink_vxlan.o tcp_metrics.o iplink_ipoib.o ipnetconf.o link_ip6tnl.o \
     link_iptnl.o link_gre6.o iplink_bond.o iplink_bond_slave.o iplink_hsr.o \
     iplink_bridge.o iplink_bridge_slave.o ipfou.o iplink_ipvlan.o \
diff --git a/ip/ip.c b/ip/ip.c
index b07a5c7..0ba29e7 100644
--- a/ip/ip.c
+++ b/ip/ip.c
@@ -68,7 +68,7 @@ static void usage(void)
 		"                   macsec | maddress | monitor | mptcp | mroute | mrule |\n"
 		"                   neighbor | neighbour | netconf | netns | nexthop | ntable |\n"
 		"                   ntbl | route | rule | sr | tap | tcpmetrics |\n"
-		"                   token | tunnel | tuntap | vrf | xfrm }\n"
+		"                   token | tunnel | tuntap | vrf | xfrm | pbbb | pbbi }\n"
 		"       OPTIONS := { -V[ersion] | -s[tatistics] | -d[etails] | -r[esolve] |\n"
 		"                    -h[uman-readable] | -iec | -j[son] | -p[retty] |\n"
 		"                    -f[amily] { inet | inet6 | mpls | bridge | link } |\n"
diff --git a/ip/iplink.c b/ip/iplink.c
index 18b2ea2..3ecd6de 100644
--- a/ip/iplink.c
+++ b/ip/iplink.c
@@ -57,7 +57,7 @@ void iplink_types_usage(void)
 		"          macsec | macvlan | macvtap |\n"
 		"          netdevsim | nlmon | rmnet | sit | team | team_slave |\n"
 		"          vcan | veth | vlan | vrf | vti | vxcan | vxlan | wwan |\n"
-		"          xfrm }\n");
+		"          xfrm | pbbb | pbbi }\n");
 }
 
 void iplink_usage(void)
diff --git a/ip/link_pbb.c b/ip/link_pbb.c
index 96e6b38..a2fd542 100644
--- a/ip/link_pbb.c
+++ b/ip/link_pbb.c
@@ -147,7 +147,7 @@ static int pbbb_parse_opt(struct link_util *lu, int argc, char **argv,
                        (n->nlmsg_flags & NLM_F_CREATE));
 	__u32 i_sid_info_range_begin = 0, i_sid_info_range_end = 0;
 	int  b_vid_info_range_begin = -1, b_vid_info_range_end = -1;
-	__u8 mode = PBB_B_MODE_MAX, key_info_map_verify = 0;
+	__u8 mode = PBB_B_MODE_MAX, key_info_map_verify = 0, key_info_map_del = 0;
 
 	fprintf(stderr, "Invoking %s with type:0x%x flags:0x%x *argv:%s\n", __FUNCTION__, n->nlmsg_type, n->nlmsg_flags, *argv);
 
@@ -184,6 +184,10 @@ static int pbbb_parse_opt(struct link_util *lu, int argc, char **argv,
 			key_info_map_verify = 1;
 
 			addattr_l(n, 1024, IFLA_PBB_B_KEY_INFO_MAP_VERIFY, &key_info_map_verify, sizeof(__u8));
+		} else if (strcmp(*argv, "key-info-map-del") == 0) {
+			key_info_map_del = 1;
+
+			addattr_l(n, 1024, IFLA_PBB_B_KEY_INFO_MAP_DEL, &key_info_map_del, sizeof(__u8));
 		} else if (matches(*argv, "i-sid") == 0) {
 			char *p = NULL;
 
@@ -265,7 +269,7 @@ static int pbbb_parse_opt(struct link_util *lu, int argc, char **argv,
 				return -1;
 		}
 
-		if ((mode == PBB_B_MODE_MAX) && (key_info_map_verify == 0)) {
+		if ((mode == PBB_B_MODE_MAX) && (key_info_map_verify == 0) && (key_info_map_del == 0)) {
 			return pbbb_b_mode_none_arg();
 		}
 
@@ -290,7 +294,7 @@ static int pbbi_parse_opt(struct link_util *lu, int argc, char **argv,
 	int c_vid_info_range_begin = -1, c_vid_info_range_end = -1;
 	int s_vid_info_range_begin = -1, s_vid_info_range_end = -1;
 	__u32 i_sid_info_range_begin = 0, i_sid_info_range_end = 0;
-	__u8 c_vid_info_action = PBB_I_VID_INFO_ACTION_MAX, s_vid_info_action = PBB_I_VID_INFO_ACTION_MAX, i_sid_type = PBB_I_ISID_TYPE_MAX, key_info_map_verify = 0;
+	__u8 c_vid_info_action = PBB_I_VID_INFO_ACTION_MAX, s_vid_info_action = PBB_I_VID_INFO_ACTION_MAX, i_sid_type = PBB_I_ISID_TYPE_MAX, key_info_map_verify = 0, key_info_map_del = 0;
 
 	fprintf(stderr, "Invoking %s with type:0x%x flags:0x%x\n", __FUNCTION__, n->nlmsg_type, n->nlmsg_flags);
 
@@ -312,6 +316,10 @@ static int pbbi_parse_opt(struct link_util *lu, int argc, char **argv,
 			key_info_map_verify = 1;
 
 			addattr_l(n, 1024, IFLA_PBB_I_KEY_INFO_MAP_VERIFY, &key_info_map_verify, sizeof(__u8));
+		} else if (strcmp(*argv, "key-info-map-del") == 0) {
+			key_info_map_del = 1;
+
+			addattr_l(n, 1024, IFLA_PBB_I_KEY_INFO_MAP_DEL, &key_info_map_del, sizeof(__u8));
 		} else if (matches(*argv, "c-vid") == 0) {
 			char *p = NULL;
 
